p4-16/psa/examples/psa-example-parser-error-handling2.p4

```diff
@@ -0,0 +1,619 @@
+    bit<16> dstPort;
+    bit<32> seqNo;
+    bit<32> ackNo;
+    bit<4>  dataOffset;
+    bit<3>  res;
+    bit<3>  ecn;
+    bit<6>  ctrl;
+    bit<16> window;
+    bit<16> checksum;
+    bit<16> urgentPtr;
+}
+
+struct empty_metadata_t {
+}
+
+struct fwd_metadata_t {
+}
+
+// BEGIN:PortId_Annotation_Example
+header to_cpu_error_header_t {
+    bit<8> error_idx;
+    bit<1> ingress;
+    bit<3> packet_path;
+    bit<4> reserved1;
+
+    // port is ingress_port if ingress is 1, else egress_port
+
+    // Note that no annotation is needed on this or any other values
+    // of type `PortIdInHeader_t` or `PortId_t`, because `psa.p4` has
+    // an annotation on the `type` definition that the compiler can
+    // propagate to all variables of these types.
+    PortIdInHeader_t port;
+}
+// END:PortId_Annotation_Example
+
+enum CloneReason_t {
+    NONE,
+    PARSER_ERROR
+    // Define other clone reasons here
+}
+
+struct clone_i2e_metadata_t {
+    CloneReason_t clone_reason;
+    to_cpu_error_header_t to_cpu_error_hdr;
+}
+
+struct clone_e2e_metadata_t {
+    CloneReason_t clone_reason;
+    to_cpu_error_header_t to_cpu_error_hdr;
+}
+
+struct metadata {
+    fwd_metadata_t fwd_metadata;
+    CloneReason_t clone_reason;
+    to_cpu_error_header_t to_cpu_error_hdr;
+}
+
+struct headers {
+    to_cpu_error_header_t to_cpu_error_hdr;
+    ethernet_t       ethernet;
+    ipv4_t           ipv4;
+    tcp_t            tcp;
+}
+
+
+control packet_path_to_bits(out bit<3> packet_path_bits,
+    in PSA_PacketPath_t packet_path)
+{
+    apply {
+        // Some P4 implementations could probably implement the below
+        // more efficiently with a table lookup, rather than a daisy
+        // chaing of if-then-elseif, as is done in the example program
+        // psa-example-parser-error-handling.p4.  In the absence of
+        // that, a switch or case statement like in C/C++/Java that
+        // could take an enum value as the expression would be
+        // equivalent to such a table, but as yet P4_16 has no such
+        // statement.
+
+        // This may be the only way as of P4_16 v1.1, PSA v1.1, and
+        // P4Runtime v1.0 to convert an enum to an integer in the data
+        // plane.
+        if (packet_path == PSA_PacketPath_t.NORMAL) {
+            packet_path_bits = 1;
+        } else if (packet_path == PSA_PacketPath_t.NORMAL_UNICAST) {
+            packet_path_bits = 2;
+        } else if (packet_path == PSA_PacketPath_t.NORMAL_MULTICAST) {
+            packet_path_bits = 3;
+        } else if (packet_path == PSA_PacketPath_t.CLONE_I2E) {
+            packet_path_bits = 4;
+        } else if (packet_path == PSA_PacketPath_t.CLONE_E2E) {
+            packet_path_bits = 5;
+        } else if (packet_path == PSA_PacketPath_t.RESUBMIT) {
+            packet_path_bits = 6;
+        } else if (packet_path == PSA_PacketPath_t.RECIRCULATE) {
+            packet_path_bits = 7;
+        } else {
+            packet_path_bits = 0;
+        }
```

Comment:
This is a clear example why we should have serializable enums. This code should have been simply
```
packet_path_bits = packet_path
```
Or am I missing something?

Reply:
We could improve on this by changing the PSA-specific `PacketPath_t` type from `enum` to a serializable `enum`, but that won't help for the `error` case.

Reply:
It won't help for the error case with the current specs, that is, because there is no "serializable error" type.

Reply:
Indeed, that's why I think we need to consider whether we want to introduce implicit casts.

Reply:
I don't think the ability to assign numerical values to errors and `PacketPath_t` members is the solution. This is really a short-coming of P4Runtime v1.0 which does not support errors & un-serializable enums as match keys.
If you support implicit casts for errors or change `PacketPath_t` to a serializable enum, you introduce more complexity for the target and potentially a lot of headaches for the programmer. As long as the numerical values are not embedded in a packet and are not used in arithmetic expressions you are fine (this is very similar to P4Runtime translation for user-defined types, for which we have explicit casts to the underlying type, but no implicit casts). But as soon as you emit the numerical value to a packet, either you are fine with a target-dependent value or the compiler itself needs to detect that situation and introduce a translation table.
I personally recommend keeping errors and un-serializable enums as they are, this is an issue that can be solved entirely in the realm of P4Runtime.

Reply:
@antoninbas I agree an explicit Cast. Working within the realm of P4Runtime, what do you think of this change for `error` and table key?  It's either bitwidth or type_error as set. Wouldn't such a change be a 1.x kind of change to P4Runtime?

```
diff --git a/proto/p4/config/v1/p4info.proto b/proto/p4/config/v1/p4info.proto
index bc95375..58823d6 100644
--- a/proto/p4/config/v1/p4info.proto
+++ b/proto/p4/config/v1/p4info.proto
@@ -170,6 +170,8 @@ message MatchField {
   Documentation doc = 6;
   // unset if not user-defined type
   P4NamedType type_name = 8;
+  // MatchField is an error
+  P4ErrorType type_error = 9;  
 }
 
 message Table {

```

Reply:
It seems to me that _if_ you are going to generalize it in that way, why not instead add a line like:
```
P4DataTypeSpec other_type = 9;
```
then say that at most one of `bitwidth` and `other_type` is set.  Then you can specify _any_ legal P4_16 type in the message.  The P4Runtime 1.x spec adding this could still limit what is supported to less than all types, e.g. maybe you do not want to jump so quickly to adding support for a header, header_union, or header stack as supported (if ever).

Reply:
`P4DataTypeSpec` uses `oneof`. Thus, we are adding a `oneof` and run into the issue below.

https://developers.google.com/protocol-buffers/docs/proto3#backwards-compatibility-issues

Reply:
I read the text at that link, and it isn't clear to me that it is recommending that you should not add a oneof.  Why do you believe it is an issue?

Reply:
The P4DataTypeSpec is included below.  The thought is "it's either bitwidth is set (non-zero) or the P4DataTypeSpec member is".  The user set P4DataTypeSpec->type_spec to bitstring.   So, do we use that fact that P4DataTypeSpec is set or because the user is using bitstring, the bitwidth is also set?  

Thus, it is best to remove bitwidth from the MatchField message - if we do, we are not backward compatible. 

```
// Describes a P4_16 type.
message P4DataTypeSpec {
  oneof type_spec {
    P4BitstringLikeTypeSpec bitstring = 1;
    P4BoolType bool = 2;
    P4TupleTypeSpec tuple = 3;
    P4NamedType struct = 4;
    P4NamedType header = 5;
    P4NamedType header_union = 6;
    P4HeaderStackTypeSpec header_stack = 7;
    P4HeaderUnionStackTypeSpec header_union_stack = 8;
    P4NamedType enum = 9;
    P4ErrorType error = 10;
    P4NamedType serializable_enum = 11;
    P4NamedType new_type = 12;
  }
}
```

Reply:
With p4runtime v1.0, the server is checking if bitwidth is set and assumes the MatchField type is `bit<W>/init<W>`.  One adds P4DataTypeSpec to p4runtime v1.1.  Now, the server checks if P4DataTypeSpec is set and determines the type.  If P4DataTypeSpec is not set, it's a failure or is the message a v1.0 message? With v1.1, bitwidth is used only to serialize the type.   I think, it's best to wait for Antonin to give his input.

Reply:
I am not here proposing the rules for a future P4Runtime spec.  I am saying that if backwards compatibility is desired, it is straightforward to achieve.  Yes, achieving backwards compatibility in the way I have suggested means that one must create a rule to specify how to interpret the future extended P4Info messages.

For example, future P4Info message creators could simply stop using `bitwidth` as a field entirely, leaving it out, using the P4DataTypeSpec message instead when the type of the table search key field is `bit<W>`.  Or alternately, the P4Runtime spec could have the rule that if the type is `bit<W>`, an implementation is free to use either the older `bitwidth` field or the newer way of doing it, and either should be accepted to have the same meaning.

To say that one _must_ break backwards compatibility by removing a field is completely against the goals of Protobuf messages.  There is no absolute need or requirement to break backwards compatibility.  The only reason to do so is some imagined convenience.